resources/assets/js/views/MenuEditor.vue

```diff
@@ -0,0 +1,400 @@
+ * 				"url": "https://kent.ac.uk"
+ * 			}
+ * 		],
+ * 		"last_published": "Fri Oct 06 2017 12:00:00 GMT+0100 (GMT Summer Time)"
+ * 	}
+ * }
+ */
+import Schema from 'async-validator';
+
+import Icon from 'components/Icon';
+import Draggable from 'vuedraggable';
+import ScrollInput from 'components/ScrollInput';
+import MenuItemField from 'components/menu-editor/MenuItemField';
+import SitePageLinks from 'components/menu-editor/SitePageLinks';
+import { win, readingSpeedFromString, prettyDate } from 'classes/helpers';
+import Config from 'classes/Config';
+
+/* global setInterval, clearInterval */
+
+export default {
+	name: 'menu-editor',
+
+	components: {
+		Draggable,
+		Icon,
+		ScrollInput,
+		SitePageLinks,
+		MenuItemField
+	},
+
+	created() {
+		this.fetchSiteData();
+
+		this.validator = new Schema({
+			text: [
+				{
+					type: 'string',
+					required: true,
+					message: 'Link text can\'t be  empty.'
+				},
+				{
+					max: 100,
+					message: 'Link text must be less than 100 characters.'
+				}
+			],
+			url: [
+				{
+					required: true,
+					message: 'Location can\'t be  empty.'
+				},
+				{
+					type: 'url',
+					message: 'Location must be a valid URL.'
+				}
+			]
+		});
+
+		this.updateTimeElapsedTimer = setInterval(
+			this.updateTimeElapsedSincePublish,
+			5000
+		);
+	},
+
+	beforeRouteLeave(to, from, next) {
+		if(this.isUnsaved) {
+			this.$confirm(
+				'Are you sure you want to leave?',
+				'There are unsaved changes',
+				{
+					confirmButtonText: 'OK',
+					cancelButtonText: 'Cancel',
+					type: 'warning'
+				}
+			).then(() => {
+				next();
+			}).catch(() => {
+				next(false);
+			});
+		}
+		else {
+			next();
+		}
+	},
+
+	beforeDestroy() {
+		clearInterval(this.updateTimeElapsedTimer);
+	},
+
+	data() {
+		return {
+			site: {
+				firstPageId: 0,
+				title: '',
+				path: ''
+			},
+
+			menu: [],
+			errors: [],
+			sitePages: [],
+
+			// serialised version of the menu, to test equality
+			// with current menu for isUnsaved computed property
+			initialMenu: null,
+			// serialised version of the published menu, to test equality
+			// with current menu for status computed property
+			publishedMenu: null,
+			// raw published date as a string
+			lastPublishedDate: null,
+			// the time elapsed since ourlast publish, in a pretty format
+			timeElapsedSincePublish: null
+		};
+	},
+
+	computed: {
+		status() {
+			return (
+				// if unsaved but is the same as published menu, treat as draft
+				JSON.stringify(this.menu) === this.publishedMenu &&
+				!this.isUnsaved ?
+					'published' : 'draft'
+			);
+		},
+
+		isUnsaved() {
+			return JSON.stringify(this.menu) !== this.initialMenu;
+		},
+
+		datePublished() {
+			const date = new Date(this.lastPublishedDate);
+			return `${date.toDateString()} at ${date.getHours()}:${date.getMinutes()}`
+		}
+	},
+
+	methods: {
+
+		fetchSiteData() {
+			this.$api
+				.get(`sites/${this.$route.params.site_id}?include=pages`)
```

Comment:
would it be useful to have a catch here? Not sure what we could usefully do other than maybe a retry? No idea...

Reply:
Yeah, I'm not sure if we should ignore unexpected/general errors like this and catch them globally or catch them whenever an API call is made. It would be cool to have some nicer mechanism for dealing with failures.